Auto-joining chat-channels

20seven opened the pull request:
Adding a setting which allows adding multiple chat-channels that will be automatically joined on client startup.
Fixes #816

micheljung commented on src/main/java/com/faforever/client/preferences/ChatPrefs.java:
```diff
@@ -46,6 +52,7 @@ public ChatPrefs() {
     userToColor = new SimpleMapProperty<>(FXCollections.observableHashMap());
     chatColorMode = new SimpleObjectProperty<>(CUSTOM);
     idleThreshold = new SimpleIntegerProperty(10);
+    autoJoinChannels = new SimpleListProperty<>(FXCollections.observableArrayList(new ArrayList<String>()));
```
`new SimpleListProperty<>(FXCollections.observableArrayList());`

micheljung commented on src/main/java/com/faforever/client/preferences/ui/AutoJoinChannelsController.java:
```diff
@@ -0,0 +1,52 @@
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class AutoJoinChannelsController implements Controller<Node> {
+  private final PreferencesService preferencesService;
+  @FXML
```
`@FXML` is only needed for private fields, by convention we use public fields w/o the annotation

micheljung commented on src/main/java/com/faforever/client/preferences/ui/AutoJoinChannelsController.java:
```diff
@@ -0,0 +1,52 @@
+  public ListView<String> channelListView;
+  public GridPane autoJoinChannelsSettingsRoot;
+
+  @Inject
```
`@Inject` is not required with only one constructor so we omit it.

micheljung commented on src/main/java/com/faforever/client/preferences/ChatPrefs.java:
```diff
@@ -164,4 +171,16 @@ public void setIdleThreshold(int idleThreshold) {
+
+  public void removeAutoJoinChannel(String channelName) {
+    autoJoinChannels.get().remove(channelName);
+  }
```
For list properties in other classes, we don't provide "add" and "remove" so, for the sake of consistency, I wouldn't do it here either.

micheljung commented on src/main/java/com/faforever/client/preferences/ui/AutoJoinChannelsController.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  public void initialize() {
+    channelListView.setOnMouseClicked(event -> {
```
By convention, we define the `onMouseClicked` method to be called in the FXML

Geosearchef reviewed (dismissed):
looks  good to me

1-alex98 committed:
Auto-joining chat-channels

Adding a setting which allows adding multiple chat-channels that will be automatically joined on client startup.
Fixes #816

Wesmania commented:
Few things to take note of:
* It's better to use getLanguage since language channels are a better fit than country channels.
* For many languages there won't really be official rooms with moderators, so I don't think there's much point in autojoining those. The Python client currently autojoins #french, #german and #russian, and these have active mods IIRC.
* Python client also has logic for autojoining the #newbie channel if current player has played few games. Maybe we can add this too.

If you'd like to look up Python client's autojoin logic for rough reference / inspiration, it's in `src/chat/channel_autojoiner.py` on develop.

20seven commented:
@Wesmania yep, language is probably better; changed it.
I also added the three "official" channels, in a way i hope is easy to add onto if needed.

The newbie channel bothered me a bit because I couldn't figure out a way that wouldn't make it all messy. I checked the server and I would consider that a much better place for this. In `lobbyconnection.py` from lines 672:

```python
channels = []
if self.player.mod:
   channels.append("#moderators")

if self.player.clan is not None:
    channels.append("#%s_clan" % self.player.clan)

 if self.player.numGames < 50:
    channels.append("#newbie")
```
I can try and explain my issues with newbie channel in slack (or maybe someone already knows a solution) but I found those two last lines much more convenient. Let me know what you think

1-alex98 commented:
@20seven any ambitions to continue here? Because otherwise I will probably redo it.
Because:
1.Now the py client is gone I would rather take the language local the user set his client and system to and join the channel with the according language code
2. I would not use a pop up but do it in the normal settings, if at all I would just join them and not leave a chance to join any other channels automatically that can be done in a separate issue

Wesmania commented:
Remember that channels need moderators, so one of the above still holds.

1-alex98 commented:
oh ok

1-alex98 commented:
Well I checked out the branch tested etc and it is perfectly fine. No idea why i wanted to redo it

1-alex98 commented:
I rebased and tested it works perfectly

1-alex98 commented:
@Geosearchef @Brutus5000 please have a look

Katharsas commented:
I cannot test this unless its rebased on top of a more recent develop commit.
Edit: I guess is can just rebase this for myself. Do you want me to push a rebased version? Otherwise ill let you deal with the merge conflicts (they probably aren't too bad) : )

1-alex98 commented:
U can happily rebase onto develop and u can also locally rebase it onto something else and not push

Wesmania commented on src/main/java/com/faforever/client/chat/OfficialLanguageChannel.java:
```diff
@@ -0,0 +1,35 @@
+
+  public static Optional<OfficialLanguageChannel> getChannelName(String localeLanguage) {
+    for (OfficialLanguageChannel channel : values()) {
+      if (channel.getLocaleLanguage().equals(localeLanguage)) {
```
Python client also adds players with 'by' locale to #russian, 'uk'  is a candidate too.

Katharsas replied:
@axel1200 Should i change the code so that it can deal with multiple locales per language channel? I don't think that would be a big change.

Katharsas commented:
Seems to work good. What about #newbie auto-joining? The following seems to work fine:
```
PlayerService playerService;
[...]
if (playerService.getCurrentPlayer().get().getNumberOfGames() < 50) {
    joinChannel("#newbie");
}
```
Edit: I pushed this change.

1-alex98 commented:
travis failed

Katharsas commented:
The more i think about it the less i like how many things the `PircBotXChatService` is doing. Maybe i can find and create a separation of concerns here.

1-alex98 commented:
Please do it separately tho.(not in this comit)

Katharsas commented:
Travis failed with some cert error.

Katharsas commented:
I need to push some additional stuff.

Katharsas committed:
Autojoin #newbie when under 50 games

Katharsas commented on src/test/java/com/faforever/client/chat/ChatControllerTest.java:
```diff
@@ -153,8 +154,7 @@ public void testOnJoinChannelButtonClicked() throws Exception {
       when(change.wasAdded()).thenReturn(true);
-      // Error here is caused by a bug in IntelliJ
-      when(change.getValueAdded()).thenReturn(new Channel(invocation.getArgument(0)));
+      doReturn(new Channel(invocation.getArgument(0))).when(change.getValueAdded());
```
A test is failing because of this. @axel1200

1-alex98 committed:
Fix unit tests

1-alex98 committed:
Fix unit tests

Brutus5000 commented on src/main/java/com/faforever/client/chat/OfficialLanguageChannel.java:
```diff
@@ -0,0 +1,35 @@
+  private final String channelName;
+
+
+  OfficialLanguageChannel(String localeLanguage, String channelName) {
```
Use Lomboks @AllArgsConstructor instead (makes the whole constructor obsolete)

Brutus5000 commented on src/main/java/com/faforever/client/chat/OfficialLanguageChannel.java:
```diff
@@ -0,0 +1,35 @@
+    return Optional.empty();
+  }
+
+  public String getChannelName() {
```
Use Lomboks @Getter on class level - makes the getter obsolete

Katharsas replied:
Im in favor of removing getters entirely. I don't see why they would be needed. static method `getChannelName` can just return the channelName directly, as its name implies.

Brutus5000 commented on src/main/java/com/faforever/client/chat/OfficialLanguageChannel.java:
```diff
@@ -0,0 +1,35 @@
+    return channelName;
+  }
+
+  public String getLocaleLanguage() {
```
Use Lomboks @Getter on class level - makes the getter obsolete

Brutus5000 commented on src/main/java/com/faforever/client/preferences/ChatPrefs.java:
```diff
@@ -48,6 +56,10 @@ public ChatPrefs() {
     chatFormat = new SimpleObjectProperty<>(ChatFormat.COMPACT);
+    autoJoinChannels = new SimpleListProperty<>(FXCollections.observableArrayList());
+    String localeLanguage = Locale.getDefault().getLanguage();
+    Optional<OfficialLanguageChannel> languageChannel = OfficialLanguageChannel.getChannelName(localeLanguage);
```
Skip the variable assignment, if it is unused and call ifPresent directly (but make a linebreak for the ifPresent).

Brutus5000 reviewed (changes requested):
Just some minor codestyle stuff.
I did not test the functionality. I hope my previous approvers did.

1-alex98 commented:
@Katharsas wanna do this ?

Katharsas commented:
Sure i think i can do this.

Katharsas committed:
Adress review and allow multiple locales per channel

Katharsas commented:
I changed the code to address the functionality brought up by @Wesmania (multiple locales) and the review by Brutus. I also removed all getters from the enum because i didn't see the point of having them.

1-alex98 commented:
Please look at https://app.codacy.com/app/FAForever/downlords-faf-client/pullRequest?prid=1492232

Katharsas committed:
Codacy

Katharsas commented:
Does selecting a language in the settings currently set my locale's language to that langauge?
For example if my local is not russian, but i set my client to russian in the settings, will `Locale.getDefault().getLanguage()` then return `ru` or mu system locale? I guess not.

Because if we use the locale from the settings instead of the default locale, i can revert the whole "multiple languages per langauge channel" thing from my last commit, because it doesn't make much sense. If people select a language other than english in the client settings then we can assume that they know that language and just put them in the appropriate channel.

This way we avoid the whole "mapping multiple languages to a single languages" thing.

Also, this whole `OfficialLanguageChannel` enum is just a disguised, hardcoded immutable map/multimap that maps channels to languages (or the other way around), and it doesn't need to an enum at all (none of its values are ever specifically in code).

Wesmania commented:
I imagine some people might want to keep an english interface (e.g. me). If we're tying joining language channels to UI language, we should have an option to specify language channels to join.

Katharsas commented:
Ok but if somebody specifically wants/selects an english interface and specifically NOT the russian one, would they actually usually want to be autojoined into the russian channel?

Btw. I don't think that the Belarusian code `by` ever worked, because the Belarusian language code is actually `be` (@Wesmania you sure that `by` was used in python?), `BY` is the country code. And do operating system even have Belarusian as a language? Is `be` actually ever returned by the operating system? If somebody sets his operating system to russian, they will be auto-joined anyway, so in that case checking for `be` would also be useless.

The whole purpose of language codes is to solve the problem of getting the language of a user correctly. For example, if an Austrian selects "Austria" as his OS language, then the operating system will return `de` (german) as the language code.
This tells me that Belarusian is actually a different language than russian, because if it was very similar, then Belarus would be `ru-BY` and not `be-BY`. So i don'''t think we should map languages to languages, because language codes should already do that by themselves.
And people can always add language channel to their settings manually to join them independently of their locale.

Katharsas commented:
I think the optimal way to deal with this would be to have an additional checkbox inside every language option in the language settings to join that language's IRC channel. By default, only the setting is ticked for the selected language. The most important thing is to SHOW that language channels exist. Even if we just show the channel name in the language settings for each language option that would be good (without checkbox).
We then ignore the OS locale completely. Kinda like this:
![language_settings](https://user-images.githubusercontent.com/1922573/61327473-495d2280-a819-11e9-8103-2cecc9e45b36.JPG)

Of course, this will mean that we can only have official language channels for languages for which we have translations, so that may be a problem. So maybe a separate UI in the chat preferecnes?

Wesmania commented:
Python client has something like that - it has an options menu for all language channels and defaults are set using system locale if not set yet.

1-alex98 commented:
@Katharsas u still wanna change it or get it done? Well tell me I would merge it now if u want to

Katharsas commented:
- I want to get rid of "multiple languages per channel" change that i made before.
- I also want to get rid of the over-testing (currently the test iterates through ALL hardcoded language channels and asserts that all of them work, which is over-testing imo), but i cannot get rid of this because i cannot mock an enum to have different values, which is an argument for changing `OfficialLanguageChannel` to be a Hashmap<String, String> instead of an enum. Should i do this?

I don't think that i want to add language channels to the chat preferences. Instead, language channels can be advertised in a potential "join new channel" UI, the same one that comes up when you close all channels. But that would belong in a different new issue.

1-alex98 commented:
Do as u please

Katharsas committed:
Moved language mapping to ChatPrefs

- Moved mapping of localeLanguages to channels to Map inside ChatPrefs, because language channels are an independent concept
- Using a proper map to get rid of all the array iteration code and the need to test the getChannelName method
- Only test one mapping instead of all mappings

Katharsas commented:
Ok so i didn't remove the multi mapping thing but since its now mapped the other way around, there is no distinction (in code) between allowing multiple locales to map to a channel or not. So i guess its fine.

Katharsas committed:
Codacy

1-alex98 commented on src/main/java/com/faforever/client/preferences/ChatPrefs.java:
```diff
@@ -21,11 +23,21 @@
 
+  @VisibleForTesting
+  public static final ImmutableMap<String, LanguageChannel> LOCALE_LANGUAGES_TO_CHANNELS = ImmutableMap.<String, LanguageChannel>builder()
+      .put("fr", FRENCH)
```
Are there not constants for that Locale.FRENCH.getLanguage() or something...

Katharsas replied:
I don't think there are.
Edit: Maybe there are.

Katharsas replied:
There are for some types. I change String type to Locale.

1-alex98 commented on src/main/java/com/faforever/client/preferences/LanguageChannel.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@AllArgsConstructor
+@Getter
```
What about @Value I thinks it makes everything final and adds getter as well as an allargs constucor

Katharsas replied:
@Value is not supported for enums, only classes.

Katharsas committed:
Use Locale type instead of string

Katharsas committed:
Fix test

1-alex98 commented on src/main/java/com/faforever/client/preferences/ChatPrefs.java:
```diff
@@ -2,24 +2,42 @@
+  public static final ImmutableMap<Locale, LanguageChannel> LOCALE_LANGUAGES_TO_CHANNELS = ImmutableMap.<Locale, LanguageChannel>builder()
+      .put(Locale.FRENCH, FRENCH)
+      .put(Locale.GERMAN, GERMAN)
+      .put(new Locale("ru"), RUSSIAN)
```
There is no constant for "ru" that seems strange
and what is "be"?

Katharsas replied:
There are no constants for most countries/languages, only for some very specific ones. `be` is Belarusian.

1-alex98 replied:
oh ok, so be it

1-alex98 commented:
it still has a ready label remove it if u dont want it merged

Katharsas commented:
I think it can be merged now.

1-alex98 commented:
lemme review one last time

Geosearchef commented:
> lemme review one last time

Review one last time in remembrance of the good old days when this thread still spammed my mail inbox every view minutes?

1-alex98 commented:
Well turn of mail notifications then they are shit anyway XD
